src/Applications/fonts/fontconfig/2.18.3/doc/fc-list.1
.\" auto-generated by docbook2man-spec from docbook-utils package
.TH "FC-LIST" "1" "Aug 13, 2008" "" ""
.SH NAME
fc-list \- list available fonts
.SH SYNOPSIS
.sp
\fBfc-list\fR [ \fB-vVh\fR ]  [ \fB--verbose\fR ]  [ \fB [ -f \fIformat\fB ]  [ --format \fIformat\fB ] \fR ]  [ \fB [ -q ]  [ --quiet ] \fR ]  [ \fB--version\fR ]  [ \fB--help\fR ] 

 [ \fB\fIpattern\fB  [ \fIelement\fB\fI...\fB ]  \fR ] 
.SH "DESCRIPTION"
.PP
\fBfc-list\fR lists fonts and styles
available on the system for applications using fontconfig.
If any elements are specified, only those are printed.
Otherwise family and style are printed, unless verbose
output is requested.
.SH "OPTIONS"
.PP
This program follows the usual GNU command line syntax,
with long options starting with two dashes (`-'). A summary of
options is included below.
.TP
\fB-v\fR
Print verbose output of the whole font pattern for each match,
or \fIelement\fRs if any is
provided.
.TP
\fB-f\fR
Format output according to the format specifier
\fIformat\fR\&.
.TP
\fB-q\fR
Suppress all normal output. returns 1 as the error code if no fonts matched.
.TP
\fB-V\fR
Show version of the program and exit.
.TP
\fB-h\fR
Show summary of options.
.TP
\fB\fIpattern\fB\fR
If this argument is set, only fonts matching
\fIpattern\fR are displayed.
.TP
\fB\fIelement\fB\fR
If set, the \fIelement\fR property
is displayed for matching fonts.
.SH "EXAMPLES"
.TP
\fBfc-list\fR
Lists all font faces.
.TP
\fBfc-list :lang=hi\fR
Lists font faces that cover Hindi.
.TP
\fBfc-list : family style file spacing\fR
Lists the filename and spacing value for each font
face. ``:'' is an empty pattern that matches all
fonts.
.SH "SEE ALSO"
.PP
\fBfc-match\fR(1)
\fBFcFontList\fR(3)
\fBFcPatternFormat\fR(3)
\fBfc-cat\fR(1)
\fBfc-cache\fR(1)
\fBfc-pattern\fR(1)
\fBfc-query\fR(1)
\fBfc-scan\fR(1)
.PP
The fontconfig user's guide <URL:https://fontconfig.pages.freedesktop.org/fontconfig/fontconfig-user.html>
.SH "AUTHOR"
.PP
This manual page was written by Keith Packard
<keithp@keithp.com> and Josselin Mouette <joss@debian.org>\&.


src/Applications/fonts/fontconfig/2.18.3/fc-list/fc-list.c
/*
 * fontconfig/fc-list/fc-list.c
 *
 * Copyright © 2002 Keith Packard
 *
 * Permission to use, copy, modify, distribute, and sell this software and its
 * documentation for any purpose is hereby granted without fee, provided that
 * the above copyright notice appear in all copies and that both that
 * copyright notice and this permission notice appear in supporting
 * documentation, and that the name of the author(s) not be used in
 * advertising or publicity pertaining to distribution of the software without
 * specific, written prior permission.  The authors make no
 * representations about the suitability of this software for any purpose.  It
 * is provided "as is" without express or implied warranty.
 *
 * THE AUTHOR(S) DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE,
 * INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS, IN NO
 * EVENT SHALL THE AUTHOR(S) BE LIABLE FOR ANY SPECIAL, INDIRECT OR
 * CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE,
 * DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER
 * TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
 * PERFORMANCE OF THIS SOFTWARE.
 */

#ifdef HAVE_CONFIG_H
#  include <config.h>
#else
#  ifdef linux
#    define HAVE_GETOPT_LONG 1
#  endif
#  define HAVE_GETOPT 1
#endif
#include <fontconfig/fontconfig.h>

#include <locale.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#ifdef HAVE_UNISTD_H
#  include <unistd.h>
#endif

#ifdef ENABLE_NLS
#  include <libintl.h>
#  define _(x) (dgettext (GETTEXT_PACKAGE, x))
#else
#  define dgettext(d, s) (s)
#  define _(x)           (x)
#endif

#ifndef HAVE_GETOPT
#  define HAVE_GETOPT 0
#endif
#ifndef HAVE_GETOPT_LONG
#  define HAVE_GETOPT_LONG 0
#endif

#if HAVE_GETOPT_LONG
#  undef _GNU_SOURCE
#  define _GNU_SOURCE
#  include <getopt.h>
const struct option longopts[] = {
    { "verbose", 0, 0, 'v' },
    { "brief",   0, 0, 'b' },
    { "format",  1, 0, 'f' },
    { "quiet",   0, 0, 'q' },
    { "version", 0, 0, 'V' },
    { "help",    0, 0, 'h' },
    { NULL,      0, 0, 0   },
};
#else
#  if HAVE_GETOPT
extern char *optarg;
extern int   optind, opterr, optopt;
#  endif
#endif

static void
usage (char *program, int error)
{
    FILE *file = error ? stderr : stdout;
#if HAVE_GETOPT_LONG
    fprintf (file, _("usage: %s [-vbqVh] [-f FORMAT] [--verbose] [--brief] [--format=FORMAT] [--quiet] [--version] [--help] [pattern] {element ...} \n"),
                     program);
#else
    fprintf (file, _("usage: %s [-vbqVh] [-f FORMAT] [pattern] {element ...} \n"),
                     program);
#endif
    fprintf (file, _("List fonts matching [pattern]\n"));
    fprintf (file, "\n");
#if HAVE_GETOPT_LONG
    fprintf (file, _("  -v, --verbose        display entire font pattern verbosely\n"));
    fprintf (file, _("  -b, --brief          display entire font pattern briefly\n"));
    fprintf (file, _("  -f, --format=FORMAT  use the given output format\n"));
    fprintf (file, _("  -q, --quiet          suppress all normal output, exit 1 if no fonts matched\n"));
    fprintf (file, _("  -V, --version        display font config version and exit\n"));
    fprintf (file, _("  -h, --help           display this help and exit\n"));
#else
    fprintf (file, _("  -v         (verbose) display entire font pattern verbosely\n"));
    fprintf (file, _("  -b         (brief)   display entire font pattern briefly\n"));
    fprintf (file, _("  -f FORMAT  (format)  use the given output format\n"));
    fprintf (file, _("  -q,        (quiet)   suppress all normal output, exit 1 if no fonts matched\n"));
    fprintf (file, _("  -V         (version) display font config version and exit\n"));
    fprintf (file, _("  -h         (help)    display this help and exit\n"));
#endif
    exit (error);
}

int
main (int argc, char **argv)
{
    int            verbose = 0;
    int            brief = 0;
    int            quiet = 0;
    const FcChar8 *format = NULL;
    FcChar8       *format_optarg = NULL;
    int            nfont = 0;
    int            i, err = 0;
    FcObjectSet   *os = 0;
    FcFontSet     *fs;
    FcPattern     *pat;
#if HAVE_GETOPT_LONG || HAVE_GETOPT
    int c;

    setlocale (LC_ALL, "");
#  if HAVE_GETOPT_LONG
    while ((c = getopt_long (argc, argv, "vbf:qVh", longopts, NULL)) != -1)
#  else
    while ((c = getopt (argc, argv, "vbf:qVh")) != -1)
#  endif
    {
	switch (c) {
	case 'v':
	    verbose = 1;
	    break;
	case 'b':
	    brief = 1;
	    break;
	case 'f':
	    format_optarg = FcStrCopy ((const FcChar8 *)optarg);
	    format = (const FcChar8 *)format_optarg;
	    break;
	case 'q':
	    quiet = 1;
	    break;
	case 'V':
	    fprintf (stderr, "fontconfig version %d.%d.%d\n",
	             FC_MAJOR, FC_MINOR, FC_REVISION);
	    exit (0);
	case 'h':
	    usage (argv[0], 0);
	default:
	    usage (argv[0], 1);
	}
    }
    i = optind;
#else
    i = 1;
#endif

    FcConfigSetWarningFlags (NULL, -1, FcTrue);
    if (argv[i]) {
	pat = FcNameParse ((FcChar8 *)argv[i]);
	if (!pat) {
	    fprintf (stderr, _("Unable to parse the pattern\n"));
	    return 1;
	}
	while (argv[++i]) {
	    if (!os)
		os = FcObjectSetCreate();
	    FcObjectSetAdd (os, argv[i]);
	}
    } else
	pat = FcPatternCreate();
    if (quiet && !os)
	os = FcObjectSetCreate();
    if (!verbose && !brief && !format && !os)
	os = FcObjectSetBuild (FC_FAMILY, FC_STYLE, FC_FILE, (char *)0);
    if (!format)
	format = (const FcChar8 *)"%{=fclist}\n";
    fs = FcFontList (0, pat, os);
    if (os)
	FcObjectSetDestroy (os);
    if (pat)
	FcPatternDestroy (pat);

    if (!quiet && fs) {
	int j;

	for (j = 0; j < fs->nfont; j++) {
	    if (verbose || brief) {
		if (brief) {
		    FcPatternDel (fs->fonts[j], FC_CHARSET);
		    FcPatternDel (fs->fonts[j], FC_LANG);
		}
		FcPatternPrint (fs->fonts[j]);
	    } else {
		FcChar8 *s;

		s = FcPatternFormat (fs->fonts[j], format);
		if (s) {
		    printf ("%s", s);
		    FcStrFree (s);
		} else {
		    err = 1;
		    break;
		}
	    }
	}
    }

    if (fs) {
	nfont = fs->nfont;
	FcFontSetDestroy (fs);
    }
    if (format_optarg) {
	free ((void *)format_optarg);
    }

    FcFini();

    return quiet ? (nfont == 0 ? 1 : err) : err;
}
